float8_experimental/float8_dynamic_linear.py

```diff
@@ -62,26 +62,26 @@ class Float8DynamicLinear(torch.nn.Linear):
     conversion to fp8 of the input and weight tensors.
     """
 
-    def __init__(self, use_activation_hooks: bool, **super_kwargs):
+    def __init__(self, cast_activation: bool, **super_kwargs):
         """
         Args:
-            use_activation_hooks (bool): whether to use activation hooks for casting to and from float8
+            cast_activation (bool): whether to do activation casting to and from float8
         """
         super().__init__(**super_kwargs)
 
-        self.use_activation_hooks = use_activation_hooks
+        self.cast_activation = cast_activation
 
     def forward(self, x):
         # cast x to float8_e4m3fn if not using activation hooks
-        x_fp8 = x if self.use_activation_hooks else self.cast_to_float8_e4m3fn(x)
+        x_fp8 = self.cast_to_float8_e4m3fn(x) if self.cast_activation else x
```

Comment:
If `self.cast_activation is False`, then what type is `x`? Would it be a `DTensor` whose local tensor is `Float8Tensor`?

I wonder if we need the bool at all. For example, the `cast_to_float8_e4m3fn` function could be idempotent in that if it is passed a `Float8Tensor`, then we return it early. The wrinkle here is that it could be a `DTensor` wrapping a `Float8Tensor`.

Reply:
I would imagine that its a DTensor(Float8Tensor) and that the only thing that changes is, do we call `Float8Tensor_constructor(DTensor(Tensor))` in the activation hook or actually in the forward

Reply:
Ohh so I guess, we cast activation is saying "DTensor is not going to handle this for you, so you need to do it?"

Reply:
if cast activation is false, it would be `DTensor(torch.Tensor[fp32/16])`, we want to turn off activation casting is because we want this casting happen inside the TP preforward/forward hooks (as it needs to happen in certain order, i.e. after `from_local` and before `redistribute`), see changes in this PR https://github.com/pytorch-labs/float8_experimental/pull/234

Reply:
updated according to our discussions :)